### ISSUE TITLE
Don't require allocation for returning string from web-sys

### ISSUE THREAD

**Issue by upsuper** [rustwasm/wasm-bindgen#873]
Currently when functions return `String` for string value from web-sys, this makes it impossible to remove allocation support from wasm when using web-sys.

I think it is possible to avoid needing to allocate. One possible approach is to define a trait like
```rust
trait StringAlloc {
    unsafe fn alloc(&mut self, len: usize) -> Option<*const u8>;
}
```
and have web-sys functions which want to return string accept a `&mut dyn StringAlloc` instead.

We can then impl that trait for `String` and have some wrapper function so people don't mind doing allocation can continue using the easy approach, while people who don't want allocation can impl that trait for their own string types like `ArrayString`.

**Comment by alexcrichton** [comment]
This seems like an interesting idea! Can you elaborate one some of the use cases you're thinking of as well? I'm not really sure how this would fit into the existing APIs as-is, but knowing more about the end goal may help!

**Comment by upsuper** [comment]
As I mentioned in the issue description, the end goal I'm thinking is to be able to have allocation-free wasm with web-sys.

For small applications, allocator is a significant part of codesize, even it's wee_alloc. (Actually in my tiny application, wee_alloc takes more codesize then the default.) So I guess it might be useful if it's possible to use web-sys without requiring allocator.

Also heap allocation is slower than pre-allocated slice in stack I suppose?

Majority of methods returning `String` are attributes in IDL, so we can probably do something like:
```rust
// binding to JS code, return whether allocation succeeded.
fn get_id(&self, &mut dyn StringAlloc) -> bool;
fn id(&self) -> String {
    let mut result = String::new();
    self.get_id(&mut result).unwrap();
    result
}
```

Oh, and it reminds me that, with doing so we can probably keep all heap allocation in the Rust side, and may no longer need to expose `__wbindgen_malloc` to JavaScript.

**Comment by alexcrichton** [comment]
This seems plausible to do but it does seem slightly esoteric in motivation. I've found that in practice `dlmalloc` is a pretty small portion of the binary and `wee_alloc` is *absolutely tiny* so if `wee_alloc` dominates the binary size it's not clear if that's the best use case for wasm. Not to say this shouldn't ever get done, it just doesn't seem too high priority to me personally at least.

**Comment by upsuper** [comment]
I agree it's probably not something very high priority. I raised this because this seems to be something doable and can allow users to remove some allocation cost when they see fit. But since web platform isn't generally considered a resource-constraint platform like embedded, this doesn't seem to be a big problem worth such effort.
